data/capture_config.yaml

```diff
@@ -0,0 +1,26 @@
+spoof_regex: .*
```

Comment:
If I'm understanding this correctly, the requested feature here is "the ability to take the command line parameters from a recent `capture start` run, and save them into a .msf4-located user config yaml file, which is used preferentially over the default in the metasploit data directory.

If I have understood that correctly, the caveat is that some settings, like whether or not certain services are on or not, are not specifiable in the command line, so it's not a silver bullet - some file modification would be required for full control over the feature. I guess we could add command line flags for that purpose, but I think that would start to get very messy to do (i.e. what would that command line look like... personally I'd rather configure it in a file as opposed to having 25 different command line settings). Especially if we start adding port numbers as well, as was suggested below.

Along the same thinking, there are some values that do not currently come from the config file (e.g. SSL Cert, log file) - for such a feature to remember these, they'd also have to come from the config file. I actually like that regardless, so I'll add that in now.

There are also some that I didn't create a command line switch for (ntlm_challenge, ntlm_domain). They seemed like less-frequent settings, and I didn't want to pollute the command line with a plethora of options, so I left them out. That means, however, that a `save` feature couldn't do anything to modify them.

The ones that I think should not come from the file would be:
- spoof_ip
- srvhost
- session
- verbose
- stdout
- help

We'd also need to change `--basic` to have a boolean flag, because otherwise you'd only be able to save it on, and never turn it back off without editing the file. Maybe `--httpmode basic` or `-httpmode ntlm` would be better in that case.

Separate question: how do you envisage the user would interact with this feature? Would they run `capture save` to save the most recently run config to the default location (excluding those that don't come from a file)? Maybe `capture save_last` is clearer?

Reply:
For me the idea that there are thing in the plugin that cannot be set via plugin commands feels like a problem in and of itself. 

The idea of having the first load of the plugin copy a default config into `~/msf4` and expecting the user to edit is a start, however diverges from how other plugins function.

Since plugins do not generated a console context like a session does the pattern has been to utilize namespaced configuration methods and offer to save those settings when desired. I would offer that pattern may server this plugin as well to offer a builder or factory pattern for configuration and usage.

See [plugins/aggregator.rb](https://github.com/rapid7/metasploit-framework/blob/60a9c8db96906dc7380498dec6943deb7a413986/plugins/aggregator.rb) and [plugins/nexpose.rb](https://github.com/rapid7/metasploit-framework/blob/60a9c8db96906dc7380498dec6943deb7a413986/plugins/nexpose.rb) as a couple examples.

Reply:
The idea of the plugin though was to remove the necessity of having to run a bunch of `set DATASTORE option` or plugin-equivalent commands. While that's been the pattern in the past, we're trying to make it easier to pass things from the command line. See the [RHOSTS URI](https://github.com/rapid7/metasploit-framework/pull/15253) changes as an example of that newer pattern.

Maybe we should just expose the missing options as parameters if that's the main concern.

Reply:
For MVP I vote to go with the file based solution for now instead of multiple `plugin_set_xyz value` commands until we've worked out what pattern folk will want to configure to suit their workflows.

Once the config file is updated to be an array of hashes (https://github.com/rapid7/metasploit-framework/pull/16298/files#r824295994 + https://github.com/rapid7/metasploit-framework/pull/16298/files#r823823074), we'll have the flexibility to persist whatever plugin configuration is required in the future, and adding extra `plugin_set_xyz` helper methods, or `capture --foo --bar` options in the future to tweak those values will be a small effort

Reply:
I'm happy to go against tool-that-shall-not-be-named's convention and use `--httpmode [basic|ntlm]`, with NTLM as a default. Shall I pull the trigger?

Reply:
For exposing the services, if we want the ability to save the last config, then we'd need to be able to say "please turn service XYZ off" as well as "please turn service XYZ on", for the same reason as the above (if you need to do the opposite of whatever the config file says so you can save a new value). Possible options:
  - `capture start ... --on http --off smb --off vnc`
  - `capture start ... --http on --smb off --vnc off`

Admittedly, I don't love this approach, though, for a few reasons:
- It's inconsistent with tool-that-shall-not-be-named, which takes the "some settings are just managed in a file; you can't command-line them" approach. I get, though, that it's probably better to be internally consistent (i.e. match MSF's behaviour rather than some other tool), so I'll hold that opinion loosely.
- As a user of this, I feel like I'd there's more mental load on me, and I'm likely to make a mistake if my defaults are pulled from a file with command line overrides. If I'm turning services off and then saving it, I'd probably be trying to hold in my head "Yeah, but which ones did I already disable?" I'd probably want to open the file to double check... and at that point, if I've got the file open anyway, I'd feel happier just editing the file since I can see it all in one place.
- Regarding @adfoster-r7's comment above about using the config file to support "multiple" services down the track (e.g. one HTTP listener on 80; another HTTP listener on 8080, defined in separate YAML entries), if we wanted to go down that path eventually (not now - I'd rather just ship what we've got, but just thinking ahead so we don't hamstring ourselves):
  - I'm not sure how a command line for an individual service would know which one to modify (e.g. if we turn "HTTP" off in a command line parameter... is that the one on 80 or the one on 8080?)
  - What about supporting alternate ports (if we eventually also allow configuring which port they are)... do we have command line parameters for them too? If so, and we support multiple HTTP listeners, how do we know which one we're referring to?
  - I realise I'm arguing against one possible feature by positing the existence of another possible feature, which is perhaps some form of madness; but I just want to make sure we're prioritising the right things for the direction we'd rather it go.

All of that having been said, happy to take direction on which approach to take. The outstanding questions for me:

- Should we allow services to be turned on/off in the command line?
- Should we implement a "save last run to .msf4" feature before shipping this?

Reply:
A few thoughts:

* Configuration file sourcing from `~/.ms4` is a **MUST**, users should never be expected to modify files that are part of the install location, this can be accomplished simply on load of the plugin by testing if the config file exists in `~/.msf4/capture_config.yml` or whatever filename is preferred and copy in the install sourced file if not found same as we do with module metadata json.
* I would prefer to see the configuration file format support the list of configured services up with initial land, documenting and changing the format later will lead to issues with support based on versions if this ships with one format and needs to change, also updating formats for users if we change over time would require some sort of migration pattern for user modified and likely end up with many edge cases.
* A possible compromise on having full configuration change options as command line methods at land could be to offer a command that opens the user's config file using in the console utilizing the default editor like the `edit` command already offers.  Something like `capture_config` could load the user's capture configuration and on save/exit of the editor update the capture configuration state. Then `capture start` and `capture stop` and `capture_config` would provide the initial functionality.  This might even be a way to avoid the need for individual command line options.
* Command line changes should not be overrides, they should impact the state of the config file. Something like `capture stop --http:8080` if/when added would update the config file then update any running services based on the config file requested state. This resolves the need to save the last state as the state is always saved.

Reply:
I've addressed those two points in the latest two commits.